### ISSUE TITLE
Add config-aware interface

### ISSUE THREAD

**Issue by XedinUnknown** [Dhii/module-interface#12]
Often, modules will want to provide a config. However, there is no way to retrieve it right now.

Add `ConfigAwareInterface`, which exposes a config in the form of `ContainerInterface`.

### Update
Hold this off for now, as it may be that there is not need for modules to expose a config in any way that is separate from other services it already exposes with `setup()`.

**Comment by mecha** [comment]
Speaking from plenty (of humble) experience with the current modular spec, separating config from services would be very inconvenient. I've encountered the following scenarios:

1) Modules wanting to access another's configs.
2) Modules wanting to override another's configs.
3) Modules providing basic config, intended for overriding or extension.
4) Modules re-using another's config as their own.
5) Modules that integrate 2 or more other modules, using their config as a guideline.

I've found that each of the above requirement can be satisfied by simply not distinguishing between config and services. This is because of the nature of container-centric DI and the current factory/extension paradigm.

In my specific cases, I've declare config as either a scalar, an associative array or even sometimes as user-land data structure. Having these directly accessible from the container makes everything incredibly easy and allows unlimited flexibility.

```php
[
	// Example of scalar config
    'db/table' => function() {
		return 'my_table';
	},
	// Example of assoc array config
    'db/info' => function (Container $c) {
		return [
			'host' => 'localhost',
			'name' => 'my_db',
			'user' => 'root',
			'pass' => '123'
		];
	},
	// Example of a service consuming config
	'db/conn' => function (Container $c) {
		$info = $c->get('db/info');

		return new mysqli($info['host'], $info['user'] /*...*/);
	},
	// Example of user-land data structure config
	'db/info' => function (Container $c) {
		$conn = $c->get('db/conn');

		return new MySqlInfoArrayObject($conn->get_server_info());
	}
]
```

As such, I think we can close this issue, unless a superior solution is found in the future.
